### PR TITLE
Updated docstring for Potential function

### PR DESCRIPTION
**Pull request by Dhruvanshu-Joshi**
**What is this PR about?**
This PR is a continuation of the PR [6489](https://github.com/pymc-devs/pymc/pull/6489) in order to solve the merge conflicts.
In this PR , I have provided a solution to close the issue [6399](https://github.com/pymc-devs/pymc/issues/6399).
The docstring for the Potential function is rather limited. I have made efforts to include some code examples to better understand this. I also focused on the nature of the arguments that are passed to this function and have included an example for the same.

**Checklist**
+ [ ] Explain important implementation details 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [ ] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Documentation
- Add code example and a short explanation of the code example to Potential docstrings.
- In the first example, we define a constraint such that the probability density that the model produces takes into consideration that x must be greater than or equal to 0.
- In the second example, we emphasis on the usage of soft constraints.
- The third example focuses on how the Potential function is used to obtain an arbitrary prior.
- In the final example, the ```soft_sum_constraint``` potential encourages x and y to have a small sum, effectively adding a soft constraint on the relationship between the two variables.

### CODE REVIEW

File: pymc/model.py [hunk 1]
```diff
@@ -2042,16 +2042,85 @@ def Deterministic(name, var, model=None, dims=None):
...
 
 def Potential(name, var, model=None):
-    """Add an arbitrary factor potential to the model likelihood
+    """
+    Add an arbitrary factor potential to the model likelihood
+
+    The Potential function is used to add arbitrary factors (such as constraints or other likelihood components) to adjust the probability density of the model.
+
+    Warnings
+    ----------
```
Comment (michaelosthege) [inline pymc/model.py]:
```suggestion
    Warnings
    --------
```
Sphinx spits out warnings if they don't align

File: pymc/model.py [hunk 2]
```diff
@@ -2042,16 +2042,85 @@ def Deterministic(name, var, model=None, dims=None):
...
+    Potential functions only influence logp based sampling, like the one used by ``pm.sample``. Potentials, modify the log-probability of the model by adding a contribution to the logp which is used by sampling algorithms which rely on the information about the observed data to generate posterior samples.
+    Potentials are not applicable in the context of forward sampling because they don't affect the prior distribution itself, only the computation of the logp.
+    Forward sampling algorithms generate sample points from the prior distribution of the model, without taking into account the likelihood function. In other words, it does not use the information about the observed data.
+    Hence, Potentials do not affect forward sampling, which is used by ``sample_prior_predictive`` and ``sample_posterior_predictive``. A warning saying "The effect of Potentials on other parameters is ignored during prior predictive sampling" is always raised to alert user of this.
```
Comment (michaelosthege) [inline pymc/model.py]:
```suggestion
    Potential functions only influence logp based sampling, like the one used by ``pm.sample``.
    Potentials, modify the log-probability of the model by adding a contribution to the logp which is used by sampling algorithms which rely on the information about the observed data to generate posterior samples.
    Potentials are not applicable in the context of forward sampling because they don't affect the prior distribution itself, only the computation of the logp.
    Forward sampling algorithms generate sample points from the prior distribution of the model, without taking into account the likelihood function.
    In other words, it does not use the information about the observed data.
    Hence, Potentials do not affect forward sampling, which is used by ``sample_prior_predictive`` and ``sample_posterior_predictive``.
    A warning saying "The effect of Potentials on other parameters is ignored during prior predictive sampling" is always emitted to alert user of this.
```
Single line breaks don't get rendered, but create simpler git diffs

File: pymc/model.py [hunk 3]
```diff
@@ -2042,16 +2042,85 @@ def Deterministic(name, var, model=None, dims=None):
...
 
     Parameters
     ----------
-    name: str
-    var: PyTensor variables
+    name : str expression
+    var : PyTensor expression
+        Expression to be added to the model joint logp
+    model : Model, optional
+        The model object to which the potential function is added. If ``None`` is provided, the current model is used.
```
Comment (michaelosthege) [inline pymc/model.py]:
```suggestion
    name : str
        Name of the potential variable to be registered in the model.
    var : tensor-like
        Expression to be added to the model joint logp.
    model : Model, optional
        The model object to which the potential function is added.
        If ``None`` is provided, the current model is used.
```

File: pymc/model.py [hunk 4]
```diff
@@ -2042,16 +2042,85 @@ def Deterministic(name, var, model=None, dims=None):
...
 
     Returns
     -------
     var: var, with name attribute
```
Comment (michaelosthege) [inline pymc/model.py]:
```suggestion
    var : tensor-like
        The registered, named model variable.
```

File: pymc/model.py [hunk 5]
```diff
@@ -2042,16 +2042,85 @@ def Deterministic(name, var, model=None, dims=None):
...
+    We pass ``-pm.math.log(pm.math.switch(constraint, 1, 0))`` as second argument which will return an expression depending on if the constraint is met or not and which will be added to the likelihood of the model.
+    The probablity density that this model produces agrees strongly with the constraint that x should be greater than or equal to 0. All the cases who do not satisfy the constraint are strictly not considered.
+
+    .. code:: python
+
+        with pm.Model() as model:
+          x = pm.Normal('x', mu=0, sigma=1)
+          y = pm.Normal('y', mu=x, sigma=1, observed=data)
+          constraint = x>= 0
+          potential = pm.Potential('x_constraint', pm.math.log(pm.math.switch(constraint, 1, 0.0)))
```
Comment (michaelosthege) [inline pymc/model.py]: The code examples should follow `black` code style, so
* 4-space indenting
* " for strings
* spaces around the `>=` operator

File: pymc/model.py [hunk 6]
```diff
@@ -2042,16 +2042,85 @@ def Deterministic(name, var, model=None, dims=None):
...
+
+    .. code:: python
+
+        with pm.Model() as model:
+          x = pm.Normal('x', mu=0.1, sigma=1)
+          y = pm.Normal('y', mu=x, sigma=1, observed=data)
+          constraint = x>=0
+          potential = pm.Potential('x_constraint', pm.math.log(pm.math.switch(constraint, 1, 0.5)))
+
+    In this example, Potential is used to obtain an arbitrary prior. This prior distribution refers to the prior knowledge that the values of max_items are likely to be small rather than being large.
```
Comment (michaelosthege) [inline pymc/model.py]:
```suggestion
    In this example, Potential is used to obtain an arbitrary prior.
    This prior distribution refers to the prior knowledge that the values of ``max_items`` are likely to be small rather than being large.
```
Here and in a few other lines you can use code formatting when mentioning variable names

**Review by michaelosthege — COMMENTED** [review]
The explanations are good!

I just nitpicked about some small code/docstring style things. You can add line breaks - at least to the level of the "1 sentence per line" rule of thumb.

**Comment by Dhruvanshu-Joshi** [comment]
Hey @michaelosthege . Thanks for the review. I have included all the suggestions in the new commit.

**Review by michaelosthege — APPROVED** [review]

File: pymc/model.py [hunk 7]
```diff
@@ -2042,16 +2042,95 @@ def Deterministic(name, var, model=None, dims=None):
...
+    Hence, Potentials do not affect forward sampling, which is used by ``sample_prior_predictive`` and ``sample_posterior_predictive``.
+    A warning saying "The effect of Potentials on other parameters is ignored during prior predictive sampling" is always emitted to alert user of this.
 
     Parameters
     ----------
-    name: str
-    var: PyTensor variables
+    name : str
+        Name of the potential variable to be registered in the model.
+    var : tensor-like
```
Comment (OriolAbril) [inline pymc/model.py]: Only issue in theprebiew is this isn't rendered as link: https://pymcio--6559.org.readthedocs.build/projects/docs/en/6559/api/generated/pymc.Potential.html#pymc.Potential.

Itshould usean underscore:

```suggestion
    var : tensor_like
```

Comment (Dhruvanshu-Joshi) [reply pymc/model.py]: Hey @OriolAbril . Updated the docstring according to this suggestion in the latest commit.

**Comment by chriswmann** [comment]
Bad timing since this was just merged but I think there are a couple of typos in the merged commit. E.g. line 2080 says "We pass `-pm.math.log(pm.math.switch(constraint, 1, 0))` as second argument" but then in the code example on line 2089, the actual argument is `pm.math.log(pm.math.switch(constraint, 1, 0.0))` (i.e. the argument is not negated). Same again for the soft constraint on lines 2091 and 2103. I'm happy to raise an issue and submit a PR if it is agreed these are corrections that should be made.

**Comment by ricardoV94** [comment]
@chriswmann thanks that would be appreciated
